SU2_CFD/include/variables/CVariable.hpp

```diff
@@ -47,2469 +47,2615 @@
 
 #include "../../../Common/include/config_structure.hpp"
 #include "../fluid_model.hpp"
+#include "../../../Common/include/toolboxes/C2DContainer.hpp"
 
 
 using namespace std;
 
+/*--- Forward declarations of some derived CVariable classes ---*/
+/*---
+ Derived CSolvers keep a private pointer of the safest type they can downcast to.
+ This bypasses most of the virtual indirection (within that solver).
+ These declarations avoid having to include all CVariable hpps in
+ solver_structure.hpp, they will not be need when that file is split.
+---*/
+class CEulerVariable;
+class CIncEulerVariable;
+class CTurbVariable;
+class CFEABoundVariable;
+class CDiscAdjVariable;
+class CDiscAdjFEAVariable;
+
 /*!
  * \class CVariable
  * \brief Main class for defining the variables.
  * \author F. Palacios
  */
 class CVariable {
 protected:
+  using Idx_t = unsigned long;
+  template<class T> using TVec_t = C2DContainer<Idx_t, T, StorageType::ColumnMajor, 64, DynamicSize, 1>;
+  template<class T> using TMat_t = C2DContainer<Idx_t, T, StorageType::RowMajor,    64, DynamicSize, DynamicSize>;
+  using Vec_t = TVec_t<su2double>;
+  using Mat_t = TMat_t<su2double>;
+
+  /*--- This contrived container is used to store matrices in a contiguous manner but still present the
+   "su2double**" interface to the outside world, it will be replaced by something more efficient. ---*/
+  struct VectorOfMatrix {
+    Vec_t storage;
+    TMat_t<su2double*> interface;
+    Idx_t M, N;
+
+    void resize(Idx_t length, Idx_t rows, Idx_t cols, su2double value) {
+      M = rows;
+      N = cols;
+      storage.resize(length*rows*cols) = value;
+      interface.resize(length,rows);
+
+      for(Idx_t i=0; i<length; ++i)
+        for(Idx_t j=0; j<rows; ++j)
+          interface(i,j) = &(*this)(i,j,0);
+    }
+
+    su2double& operator() (Idx_t i, Idx_t j, Idx_t k) { return storage(i*M*N + j*N + k); }
+    const su2double& operator() (Idx_t i, Idx_t j, Idx_t k) const { return storage(i*M*N + j*N + k); }
+
+    su2double** operator[] (Idx_t i) { return interface[i]; }
+  };
+
+  Mat_t Solution;       /*!< \brief Solution of the problem. */
+  Mat_t Solution_Old;   /*!< \brief Old solution of the problem R-K. */
```

Comment:
Some data structures have been added (mainly computational graph references/indices) on the variable level in #774 with some subsequent changes for the routines building on top of them. The changes here should go first so that one can apapt them easier.

Reply:
I prefer if we do it the other way, merge every PR with significant changes to CVariable, solver etc. ahead of this one, I'll fix the conflicts and port what needs porting.
I have some scripts to make my live easier, don't worry.

Reply:
Ok, that's important to know!